src/main/scala/datalog/execution/ir/IRTreeGenerator.scala

```diff
@@ -123,16 +134,25 @@ class IRTreeGenerator(using val ctx: InterpreterContext)(using JITOptions) {
   def generateSemiNaive(ast: ASTNode): IROp[Any] = {
     ast match {
       case ProgramNode(ruleMap) =>
-        ProgramOp(SequenceOp(OpCode.SEQ,
-          naiveEval(ruleMap, true),
-          DoWhileOp(
-            DB.Delta,
-            SequenceOp(OpCode.LOOP_BODY,
-              SwapAndClearOp(),
-              semiNaiveEval(ctx.toSolve, ruleMap)
-            )
-          )
-        ))
+        val scc = ctx.precedenceGraph.scc(ctx.toSolve)
```

Comment:
Same comment here, this should be `generateStratifiedSemiNaive` that calls into the untouched `generateSemiNaive` method.